runtime/server/x86/frontend/wav.h

```diff
@@ -0,0 +1,188 @@
+  char riff[4];  // "riff"
+  unsigned int size;
+  char wav[4];  // "WAVE"
+  char fmt[4];  // "fmt "
+  unsigned int fmt_size;
+  uint16_t format;
+  uint16_t channels;
+  unsigned int sample_rate;
+  unsigned int bytes_per_second;
+  uint16_t block_size;
+  uint16_t bit;
+  char data[4];  // "data"
+  unsigned int data_size;
+};
+
+class WavReader {
+ public:
+  explicit WavReader(const std::string& filename) {
+    FILE* fp = fopen(filename.c_str(), "r");
+    if (NULL == fp) {
+      perror(filename.c_str());
+      exit(1);
+    }
+
+    WavHeader header;
+    fread(&header, 1, sizeof(header), fp);
+    if (header.fmt_size < 16) {
+      fprintf(stderr, "WaveData: expect PCM format data "
+                      "to have fmt chunk of at least size 16.\n");
+      exit(1);
+    } else if (header.fmt_size > 16) {
+      int offset = 44 - 8 + header.fmt_size - 16;
+      fseek(fp, offset, SEEK_SET);
+      fread(header.data, 8, sizeof(char), fp);
+    }
+    // check "riff" "WAVE" "fmt " "data"
+    // only support one sub data chunk
+    num_channel_ = header.channels;
+    sample_rate_ = header.sample_rate;
+    bits_per_sample_ = header.bit;
+    int num_data = header.data_size / (bits_per_sample_ / 8);
+    data_ = new float[num_data];
+    num_sample_ = num_data / num_channel_;
+
+    for (int i = 0; i < num_data; i++) {
+      switch (bits_per_sample_) {
+        case 8: {
+          char sample;
+          fread(&sample, 1, sizeof(char), fp);
+          data_[i] = static_cast<float>(sample);
+          break;
+        }
+        case 16: {
+          int16_t sample;
+          fread(&sample, 1, sizeof(int16_t), fp);
+          data_[i] = static_cast<float>(sample);
+          break;
+        }
+        case 32: {
+          int sample;
+          fread(&sample, 1, sizeof(int), fp);
+          data_[i] = static_cast<float>(sample);
+          break;
+        }
+        default:
+          fprintf(stderr, "unsupported quantization bits");
+          exit(1);
+      }
+    }
+    fclose(fp);
+  }
+
+  int NumChannel() const { return num_channel_; }
+  int SampleRate() const { return sample_rate_; }
+  int BitsPerSample() const { return bits_per_sample_; }
+  int NumSample() const { return num_sample_; }
+
+  ~WavReader() {
+    if (data_ != NULL) delete data_;
+  }
+
+  const float *Data() const { return data_; }
+
+ private:
+  int num_channel_;
+  int sample_rate_;
+  int bits_per_sample_;
+  int num_sample_;  // sample points per channel
+  float *data_;
+};
+
+class WavWriter {
+ public:
+  WavWriter(const float* data, int num_sample, int num_channel,
+            int sample_rate, int bits_per_sample):
+    data_(data), num_sample_(num_sample),
+    num_channel_(num_channel),
+    sample_rate_(sample_rate),
+    bits_per_sample_(bits_per_sample) {}
+
+  void Write(const std::string& filename) {
+    FILE* fp = fopen(filename.c_str(), "w");
+    // init char 'riff' 'WAVE' 'fmt ' 'data'
+    WavHeader header;
+    char wav_header[44] = {
```

Comment:
Documentation is missing

Reply:
this is also a tmp wav file, will refine it later.